# formulas
- area of circle=pi*r**2
- area of triangle=0.5*breadth*height
- perimeter of rectangle=2*(l+b)
- area of rectangle=l*b
- volume of the cylinder=pi*r*r*h

In [35]:
0b101

5

In [3]:
0B101

5

In [4]:
0O123

83

In [5]:
0o123

83

In [6]:
0xa1c

2588

In [7]:
0XA1c

2588

In [8]:
1e+3

1000.0

In [9]:
1e+1,1e+2,1e+3

(10.0, 100.0, 1000.0)

In [10]:
1e-1,1e-2,1e-3,1e-4

(0.1, 0.01, 0.001, 0.0001)

# type casting-type converstion means type casting

## int to all datatypes

In [16]:
bool(100),bool(-100),bool(0)

(True, True, False)

## float to all datatypes

In [12]:
bool(0.0)

False

## string to all datatypes
## three cases

In [14]:
str('100.23')

'100.23'

- case_1

In [19]:
print(int('100')) #true
print(float('100'))#true
print(bool('100'))#true

100
100.0
True


- case_2

In [20]:
int('100.23')
#float is boss
#integer converstion float value in string format will fail
#integer will accept the float value   ex: int(100.25)  works
#integer will not accept the float value in strings format  ex: int('100.25') fail

ValueError: invalid literal for int() with base 10: '100.23'

- case-3

In [22]:
int('python')   # Fail
float('python') # Fail
bool('python')  # True

ValueError: invalid literal for int() with base 10: 'python'

## boolean to all other datatypes

In [27]:
bool('100')  # True
bool('100.25') # True
bool('0')    # True
bool('0.0')  # True
bool('python') # True
bool(100)  # T
bool(-100) # T
bool(100.25) # T
bool(0)  # F   (Off )
bool(0.0) # F  (off)
bool('')  # F   (off)

False

# print statement

In [28]:
a=10
b=20
print(add=a+b)

TypeError: 'add' is an invalid keyword argument for print()

In [29]:
num1=200
num2=30
add=num1+num2
print("the addition of",num1,'and',num2,'is',add)

the addition of 200 and 30 is 230


# format

In [31]:
#method_1
a=10
b=20
add=a+b
print("the addition is {} and {} is {}".format(a,b,add))

the addition is 10 and 20 is 30


## f_string method_2

In [32]:
a=10
b=20
add=a+b
print(f"the addition is {a} and {b} is {add}")

the addition is 10 and 20 is 30


## round

In [34]:
a=123.45678
round(a,3)

123.457

## end
- this is for if print statement has one value or one statement

In [36]:
print('hello',end=' ')  # 1
print('good morning') # 2  

hello good morning


In [39]:
print(1,end='@')
print(2,end='$')
print(3,end="#")

1@2$3#

## sep
- this is for if print statement have more than one value or multiple value

In [2]:
print(28,6,2001,sep="-")

28-6-2001


In [4]:
print('a','b','c',sep=",")

a,b,c


## input

In [7]:
n=input("Enter the string")
k=int(input("Enter the integer"))
j=float(input("Enter the float"))
print(n)
print(k)
print(j)

Enter the string krishna
Enter the integer 6
Enter the float 7


krishna
6
7.0


# three ways to provide the number
- hard coded
          - a=10
- from keyboard
          - a=eval(input())
- using random package
          - a=random.randint(1,100)

## eval
- eval means evaluate
- evalute will use for numbers only, not for english
- evaluate will convert integer number to integer
- evaluate will convert float numbers to float

In [11]:
n=eval(input("Enter the number"))
j=eval(input("Enter the float"))
print(n)
print(j)

Enter the number 7
Enter the float 8.0


7
8.0


In [16]:
# Q1) Wap ask the user enter two numbers from keyboard
# print addition
# print subtraction
# print multiplication
# print division using f string method

a=eval(input("Enter the first number: "))
b=eval(input("Enter the second number: "))
A=a+b
s=a-b
m=a*b
d=a/b
print(f"addition of {a} and {b} is : {A}")
print(f"subtraction of {a} and {b} is : {s}")
print(f"multiplication of {a} and {b} is : {m}")
print(f"division of {a} and {b} is : {d}")

Enter the first number:  4
Enter the second number:  3


addition of 4 and 3 is : 7
subtraction of 4 and 3 is : 1
multiplication of 4 and 3 is : 12
division of 4 and 3 is : 1.3333333333333333


In [17]:
# Q2) Ask the user enter 3 numbers and find the average of those numbers
a=int(input())
b=int(input())
c=int(input())
avg=(a+b+c)//3
print(f"average of {a} {b} {c} is: {avg}")

 8
 7
 6


average of 8 7 6 is: 7


In [18]:
 #Q3)  Ask the enter the bill
      # ask the user ho much tip percentage
      # calculate the total bill

# 1000 rs
# 10% : 100
# 1000+100
bill=int(input())
tip=int(input())
tip_amount=(bill*tip)/100
total=bill+tip_amount
print(total)

 1000
 10


1100.0


In [22]:
#Q4) ask the enter the weight in kgs
#    display the answer in pounds
#    1kg=2.2 pounds
weight=int(input())
pound=round(weight*2.2,3)
print(pound)

 100


220.0


## random

In [26]:
import random
k=random.randint(10,30)
j=random.randint(70,90)
add=k+j
print(f"addition of random numbers {k} and {j} is: {add}")

addition of random numbers 24 and 77 is: 101


# Error
- syntax error

  - if you miss the ) , "", :

- name error

    - if you not defined any name that in black color

- indentation error

    - after : we need some space

- division error

    - some number divided by zero

- value error

    - number type conversion

- type error

    - any math operations on engish letters

## 1.Name Error

In [33]:
weight_kg=eval(input("Enter the weight in kgs: "))
pound=eval(input("how many pounds for 1 kg:"))
weight_pound=round(poundk*weight_kg,2)
print(f"the weight in pound is {weight_pound}")

Enter the weight in kgs:  5
how many pounds for 1 kg: 6


NameError: name 'poundk' is not defined

## 2.Value Error

In [57]:
int('100.23')

ValueError: invalid literal for int() with base 10: '100.23'

## 3.Type Error

In [58]:
a=10
b=20
print(add=a+b)

TypeError: 'add' is an invalid keyword argument for print()

In [8]:
n1=input("enter a num1:")
n2=input("enter a num1:")
n1*n2

enter a num1: krishna
enter a num1: veni


TypeError: can't multiply sequence by non-int of type 'str'

## 4.Indentation Error

In [60]:
if 100>10:
print("manish")

IndentationError: expected an indented block after 'if' statement on line 1 (1658022112.py, line 2)

## 5.syntax Error

In [62]:
if 100>10
    print("manish")

SyntaxError: expected ':' (4054628281.py, line 1)

## 6.ZeroDivisionError

In [74]:
print(1)
if 100>10:
    print("good condition is correct")
    print(2/0)

else :
    print("Bad condition is wrong")
    print(4)
print(5)

1
good condition is correct


ZeroDivisionError: division by zero

# time
- time.sleep(2)

In [34]:
# Print all 4 lines 2 seconds gap
import time
n1=eval(input("enter the number1:"))
n2=eval(input("enter the number2:"))
add=n1+n2
sub=n1-n2
mul=n1*n2
div=round(n1/n2,2)
print(f"the addition of {n1} and {n2} is {add}")
time.sleep(2)
print(f"the subtraction of {n1} and {n2} is {sub}")
time.sleep(2)
print(f"the multiplication of {n1} and {n2} is {mul}")
time.sleep(2)
print(f"the division of {n1} and {n2} is {div}")

enter the number1: 4
enter the number2: 2


the addition of 4 and 2 is 6
the subtraction of 4 and 2 is 2
the multiplication of 4 and 2 is 8
the division of 4 and 2 is 2.0


# math

In [75]:
# area of circle
import math
pi=math.pi
r=eval(input("enter the radius:"))
area=round(pi*r*r,2)
print(area)

enter the radius: 20


1256.64


In [42]:
#area of triangle
b=int(input())
h=int(input())
area_of_triangle=0.5*b*h
print(area_of_triangle)

 20
 10


100.0


In [47]:
# by three ways
#perimeter of rectangle hard coded
l=14
b=20
pr=2*(l+b)
print(pr)

68


In [48]:
#area of rectangle key board
l=eval(input())
b=eval(input())
ar=l*b
print(ar)

 20
 20


400


In [50]:
#perimeter of rectangle
#area of rectangle
# by random package method
import random
l=random.randint(10,30)
b=random.randint(10,30)
area_of_rectangle=l*b
perimeter_of_rectangle=2*(l+b)
print(f"area of rectangle {l} and {b} is: {area_of_rectangle}")
print(f"perimeter of rectangle {l} and {b} is: {perimeter_of_rectangle}")

area of rectangle 30 and 19 is: 570
perimeter of rectangle 30 and 19 is: 98


In [56]:
#wap ask the user calculate volume of the cylinder
# Formulae: pi*r*r*h
import random
import math
pi=math.pi
r=random.randint(10,20)
h=random.randint(20,30)
volume_of_cylinder=round(pi*r*r*h,2)
print(f"volume of cylinder {r} and {h} is: {volume_of_cylinder}")

volume of cylinder 12 and 22 is: 9952.57


# conditional statements
- if
- if-else
- if-elif-else

## if condition is used to print statement while the condition is true

In [63]:
if True:
    print("python")

python


In [64]:
if False:
    print("python")

In [65]:
if 10<100:
    print("10 is smallerthan 100")

10 is smallerthan 100


In [67]:
print(1)    
print(2)    
if 100<10: #False
    print("manish")
    print("bye")
print("hello") 
print("python") 

1
2
hello
python


## if-else condition is used to print statement while the if condition is false then else bolck print the statement
- else block does not require condition

In [71]:
if 100<10:
    print("good condition is correct")
else:
    print("Bad condition is wrong")

Bad condition is wrong


In [73]:
print(1)
if 100<10:
    print("good condition is correct")
    print(2)

else :
    print("Bad condition is wrong")
    print(4)
print(5)

1
Bad condition is wrong
4
5


In [76]:
#WAP ask the user enter a value from keyboard
# Find whether the number is even or odd

# hint: num%2==0
# step-1: get the number from keyboard
# step-2: if <condition>:
# step-3:    print(f"the {num} is even")
# step-4: else:
# step-5:    print(f"the {num} is odd")

# = means assigning a vlaue
# == checking the condition
num=eval(input("enter the number:"))
if num%2==0:
    print(f"the {num} is even")
else:
    print(f"the {num} is odd")

enter the number: 6


the 6 is even


In [79]:
# wap implment above the code 
# by taking number randomly between 10,100
import random
num=random.randint(10,100)
if num%2==0:
    print(f"the {num} is even")
else:
    print(f"the {num} is odd")

the 71 is odd


In [81]:
# Game program
# there are two numbers
# num1 comes from random
# num2 is taking from the keyboard
# if both numbers are equal 
#     then print  you won
# else 
#       print you loss
import random
num1=random.randint(1,10)
print(num1)
num2=eval(input("enter the number:"))
if num1==num2:
    print("won")
else:
    print("loss")

4


enter the number: 7


loss


## if-elif-else condition is used when we have more then one condition
- here first condition refer to if block
- last condition under else block
- remaining all other conditions under elif block

In [83]:
num = eval(input("enter a number"))
if num==1:
    print(f"{num} is mach")
elif num==2:
    print(f"{num} is mach")
elif num==3:
    print(f"{num} is mach")
elif num==4:
    print(f"{num} is mach")
else:
    print("number not mach")

enter a number 5


number not mach


In [85]:
# wap ask the user
# enter marks percentage
# If percentage greater than 90 then print A grade
# If percentage between 70 and 90 then print B grade
# If percentage between 50 and 70 then print C grade
# If percentage between 35 and 50 then print D grade
# otherwise print fail
percentage=int(input("Enter the percentage: "))
if percentage>=90:
    print("A grade")
elif percentage>=70 and percentage<90:
    print("B grade")
elif percentage>=50 and percentage<70:
    print("C grade")
elif percentage>=35 and percentage<50:
    print("D grade")
else:
    print("fail")
    

Enter the percentage:  89


B grade


In [88]:
# add
# mul 
# sub
# div based on conditions

# take two numbers 
# num1 from keyboard
# num2 from keyboard

# enter the operation 1 for addition
# enter the operation 2 for mul
# enter the operation 3 for sub
# enter the operation 4 for div

# opertaion=ask the user enter opertaion from keyboard
# if operation equal to 1
#     perform the addition
# if operation equal to 2
#     perform the sub
# if operation equal to 3
#     perform the mul
# if operation equal to 4
#     perform the div

num1=eval(input("Enter the number : "))
num2=eval(input("Enter the number : "))

print("Enter the operation 1 for addition")
print("Enter the operation 2 for subtraction")
print("Enter the operation 3 for multiplication")
print("Enter the operation 4 for division")

operation=int(input("Enter the number from 1 to 4 : "))
if operation==1:
    print(num1+num2)
elif operation==2:
    print(num1*num2)
elif operation==3:
    print(num1-num2)
elif operation==4:
    print(num1/num2)

Enter the number :  4
Enter the number :  6


Enter the operation 1 for addition
Enter the operation 2 for subtraction
Enter the operation 3 for multiplication
Enter the operation 4 for division


Enter the number from 1 to 4 :  3


-2


In [90]:
#wap
# ask the user enter gender
# if gender equal to male
#     ask the user enter age
#     if age greater tha 30 then print middilaged men
#     otherwise print boy
# elif gender equal to female
#     ask the user enter age
#     if age greater tha 30 then print middilaged woman
#     otherwise print girl
# othrwise
#     print enter valid gender

gender=input("Enter the gender : ")
if gender=="male":
    age=eval(input("Enter the age : "))
    if age>30:
        print("middilaged men")
    else:
        print("boy")
elif gender=="female":
    age=eval(input("Enter the age : "))
    if age>30:
        print("middilaged woman")
    else:
        print("girl")
else:
    print("Enter valid gender")

Enter the gender :  female
Enter the age :  23


girl


In [91]:
#wap ask the user enter 3 numbers
# find the biggest numbers
a=eval(input("Enter the number: "))
b=eval(input("Enter the number: "))
c=eval(input("Enter the number: "))
if a>b and a>c:
    print("a is the biggest number")
elif b>a and b>c:
    print("b is the biggest number")
else:
    print("c is the biggest number")

Enter the number:  6
Enter the number:  9
Enter the number:  5


b is the biggest number


# try-except

- Generally python is a sequencital process

- which mean if you get error, the compiler will not execute other statements

- There is a situations even though if you got error at particular lines

- Compiler should execute some part of the code

- This is possibile by using try-except method

- we have two blocks  

    - try block

        - original code will run under try block

    - except block

        - if any error comes it will redirect to except block
        - syntax errors will not capture by exception block

In [1]:
#Example
a=10
b=0
a/b
print("hello")

ZeroDivisionError: division by zero

**case1**
- whenever we got the error in try block

- the comipler will execute statements of except block

In [4]:
try:
    a=10
    b=0
    print("the value a is:",a)
    a/b
    print("you will get error")
except:
    print("hello")

the value a is: 10
hello


**case-2**

- If there is no error in the try block

- the comipler will not execute statements of except block

In [5]:
try:
    a=10
    b=10
    print("the value a is:",a)
    print("the value b is:",b)
    print(a/b)
    print("you will not get error")
except:
    print("hello")

the value a is: 10
the value b is: 10
1.0
you will not get error


**Capture the errors**
- Mainly try and except is used to capture the errors

- syntax error

  - if you miss the ) , "", :

- name error

    - if you not defined any name that in black color

- indentation error

    - after : we need some space

- division error

    - some number divided by zero

- value error

    - number type conversion

- type error

    - any math operations on engish letters

In [16]:
import random
try:
    num=random.randint(10,30)
    if num%2==0:
        print(f"{num} is even")
    else:
        print(odd)
except:
    print(f"{num} enter valid string")
    

10 is even


# Functions : function also called a method

- Reuse of the code

- Imagine you want to calculate tax of all tax payers

- Suppose there are 100 tax payers are there

- In order to calculate tax we required 5 lines of code

- for 100 tax payers it will required 500 lines

- this is not the good appraoch

- because 5lines we are repaeting for 100 members

- Here Functions will help us how to use only these 5 lines to all 100 tax payers
  
**Mistakes**

- first mistake you will forget the brackets while define function

  - syntax error

- second mistake you will forget colon

  - syntax error

- third mistake you will not give indetation

    - indetation error

- Fourth mistake, you will copy the code you will forget to make the allignment

    - indetation error

- fifth mistake: variable name and functions name must be different

    - This error you will get in future, when we create apps

- sixt mistake: while calling the function you will
 
**Note**

- when we define function will not give the error

- erros will display after calling the function only forget the brackets

In [17]:
#Example
salary=eval(input('enter the salary:'))
tax_per=eval(input('enter the tax per:'))
tax_amount= salary*tax_per/100
print(f"the amount of tax pay is {tax_amount}")

enter the salary: 60000
enter the tax per: 10


the amount of tax pay is 6000.0


In [25]:
def tax(a,b):
    tax_amount=salary*tax_per/100
    return tax_amount
salary=int(input("Enter the salary: "))
tax_per=int(input("Enter the tax_per: "))
tax_payer=tax(salary,tax_per)
print(tax_payer)


Enter the salary:  60000
Enter the tax_per:  10


6000.0


## Method-1 :function without argument

- when we define function we will provide brackets

- inside these brackets we can provide the values also

- these values are called as **arguments** or **parameters**

- if we dont provide argumnets it is known as

  - **function with out arguments**

In [29]:
def average():
    n1=int(input())
    n2=int(input())
    n3=int(input())
    avg=(n1+n2+n3)/3
    print(avg)
average()

 3
 4
 5


4.0


## Method-2 : function with arguments

- how many variables are there in a given function

    - input variables : user will provide
 
    - suppose n1 and n2 are two input variables
 
    - output variables: we are creating a new varible using input variable
 
**concentrate on input variables**

In [31]:
import random
def billing(bill,tip_per):
    print("the bill is:",bill)
    print("the tip per is:",tip_per)
    tip_amount=(bill*tip_per)/100
    total=bill+tip_amount
    print(total)

billing(random.randint(1000,2000),
        random.randint(10,20))

the bill is: 1240
the tip per is: 11
1376.4


## Method-2 : Default argument

- argument values are fixed

In [33]:
def billing(bill,tip_per=20):
    print("the bill is:",bill)
    print("the tip per is:",tip_per)
    tip_amount=(bill*tip_per)/100
    total=bill+tip_amount
    print(total)

billing(1000)

the bill is: 1000
the tip per is: 20
1200.0


**case_1**
- all the arguments are default

- so no need to pass the value while calling the function

In [36]:
def even(num=10):
    if num%2==0:
        print("even")
    else:
        print("odd")
even()

even


**Case-2**
- All default parameters should assign after non default arguments

In [38]:
def avg(num1,num2,num3=10):
    average=(num1+num2+num3)/3
    print(average)
avg(2,3)

5.0


**Case-3**
- Default argumnents values will be override

In [39]:
def billing(bill,tip_per=20):
    print("the bill is:",bill)
    print("the tip per is:",tip_per)
    tip_amount=(bill*tip_per)/100
    total=bill+tip_amount
    print(total)

billing(1000,40)

the bill is: 1000
the tip per is: 40
1400.0


In [40]:
def billing(bill=1000,tip_per=20):
    bill=5000
    print("the bill is:",bill)
    print("the tip per is:",tip_per)
    tip_amount=(bill*tip_per)/100
    total=bill+tip_amount
    print(total)

billing(2000,40)

the bill is: 5000
the tip per is: 40
7000.0


# Global variable - Local variable

- local variable means the variables define inside the function

- local variables can not access outside the function

- global variables means the variables define outside the function

- global variable can access anywhere anytime


In [2]:
#example for local variales
def addition():
    NUMBER1=10
    NUMBER2=20
    print(NUMBER1+NUMBER2)

addition()

30


In [8]:
#It will give error because local variables can not access outside the function
NUMBER1

NameError: name 'NUMBER1' is not defined

In [5]:
#Example for Global variables

In [6]:
number1=10
number2=20
def addition(a,b):
    add=a+b
    print(add)
addition(number1,number2)

30


In [11]:
#It will not give error and it will giv output as value because global variables can  access outside the function
number2

20

In [12]:
NUMBER11=100
NUMBER22=200
def addition():
    print("number1:",NUMBER11)
    print("number2:",NUMBER22)
    add=NUMBER11+NUMBER22
    print(add)
NUMBER11=1000
NUMBER22=2000
addition()

number1: 1000
number2: 2000
3000


## spl case
- we want to use the local variable outside function also
  - i defined variable as global inside function
  -  with out call the function als
  - i can use that variable anywherere

In [17]:
def addition():
    global number_one1
    number_one1=10
    number_two=20
    print(number_one1+number_two)
addition()
print(number_one1)

30
10


In [18]:
number_one1

10

## return
- We can also use the local variables with out using global keyword

- that concept is called **return** 

In [24]:
def add1():
    num1=10
    num2=20
    summ=num1+num2
    return(summ)

var1=add1()

In [25]:
var1

30

In [26]:
def add1():
    num1=10
    num2=20
    s11=num1+num2
    return(s11,num1,num2)

s11,num1,num2=add1()
print(s11)
print(num1)
print(num2)

30
10
20


# Function in Function

In [1]:
try:
    def krishna():
        def addition(a,b):
            add=a+b
            print(add)
        def subtraction(a,b):
            sub=a-b
            print(sub)
        def multiplication(a,b):
            mul=a*b
            print(mul)
        def division(a,b):
            div=a/b
            print(div)
        print("Enter operation=1 then perform addition operation")
        print("Enter operation=2 then perform sub operation")
        print("Enter operation=3 then perform mul operation")
        print("Enter operation=4 then perform div operation")
        operation=int(input("Enter the operation :"))
        if operation==1:
            addition(2,3)
        elif operation==2:
            subtraction(4,2)
        elif operation==3:
            multiplication(2,3)
        elif operation==4:
            division(10,5)
    repetion=input()
    if repetion=="yes":
        krishna()
    else:
        print("No repetion")
        
except:
    if repetion !="yes" :
        print("Give repetion as yes")

 yes


Enter operation=1 then perform addition operation
Enter operation=2 then perform sub operation
Enter operation=3 then perform mul operation
Enter operation=4 then perform div operation


Enter the operation : 1


5


# Loops
- Loop:  Reuse the code for multiple times unless the given conditions satiesfies

  - For
 
  - While
 
- Any loop requires 3 things

  - start point
 
  - what is the condition to stop
 
  - Increment or decrement
 
- in for loop we will apply all above 3 in a single line

- in while loop we will apply all above 3 in 3 different lines

## For loop
**method-1**

In [4]:
for i in range(5):
    print(i)

0
1
2
3
4


**method-2**

In [6]:
for i in range(2,6):
    print(i)

2
3
4
5


In [9]:
import random
for i in range(5):
    user=random.randint(10,100)
    print(f"{user} is {user**2}")

11 is 121
97 is 9409
14 is 196
88 is 7744
60 is 3600


In [12]:
def square():
    user=random.randint(10,100)
    return(f"{user} is {user**2}")
for i in range(5):
    a=square()
    print(a)

64 is 4096
86 is 7396
48 is 2304
59 is 3481
82 is 6724


**method-3**

In [17]:
#case-1
for i in range(2,20,2):
    print(i, end=" ")

2 4 6 8 10 12 14 16 18 

In [18]:
#case-2
for i in range(2,20,-2):
    print(i)

In [19]:
#case-3
for i in range(2,-20,-2):
    print(i,end=" ")

2 0 -2 -4 -6 -8 -10 -12 -14 -16 -18 

In [20]:
#case-4
for i in range(-2,-20,-2):
    print(i,end=" ")

-2 -4 -6 -8 -10 -12 -14 -16 -18 

In [21]:
#case-5
for i in range(-2,20,-2):
    print(i)

In [23]:
 for i in range(19,3,-3):
     print(i,end=" ")

19 16 13 10 7 4 

**wrapper**
- Counter wrapper

In [25]:
summ=0
for i in range(1,10):
    summ=summ+i
print(summ)

45


# keywords

In [26]:
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

# operator in for loop
- in operator -charcters
- range operator -numbers

In [27]:
name="Python"
for i in name:
    print(i)

P
y
t
h
o
n


# ASCII - American standard Code for information interchange
- space -32
- special characters -33 to 47
- digits(0-9) -48 to 57
- arthemativc symbols-58 to 64
- (A-Z) 65 to 90
- bracket -91 to 96
- (a-z) -97 to 122
- { ~(123 to 126)

**ord-chr**

- ord is the keyword which provides ASCII number of Charecters

- char is the keyword which provides charecters for given ASCII number

In [29]:
ord('A'), ord('a')

(65, 97)

In [30]:
for i in "ABCDEFGHIJKLMNOPQRSTUVWXYZ":
    print(i,ord(i),end=" ")

A 65 B 66 C 67 D 68 E 69 F 70 G 71 H 72 I 73 J 74 K 75 L 76 M 77 N 78 O 79 P 80 Q 81 R 82 S 83 T 84 U 85 V 86 W 87 X 88 Y 89 Z 90 

In [31]:
import string
dir(string)

['Formatter',
 'Template',
 '_ChainMap',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_re',
 '_sentinel_dict',
 '_string',
 'ascii_letters',
 'ascii_lowercase',
 'ascii_uppercase',
 'capwords',
 'digits',
 'hexdigits',
 'octdigits',
 'printable',
 'punctuation',
 'whitespace']

In [32]:
string.ascii_letters

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [34]:
for i in range(32,127):
    print(i,chr(i),end=" ")

32   33 ! 34 " 35 # 36 $ 37 % 38 & 39 ' 40 ( 41 ) 42 * 43 + 44 , 45 - 46 . 47 / 48 0 49 1 50 2 51 3 52 4 53 5 54 6 55 7 56 8 57 9 58 : 59 ; 60 < 61 = 62 > 63 ? 64 @ 65 A 66 B 67 C 68 D 69 E 70 F 71 G 72 H 73 I 74 J 75 K 76 L 77 M 78 N 79 O 80 P 81 Q 82 R 83 S 84 T 85 U 86 V 87 W 88 X 89 Y 90 Z 91 [ 92 \ 93 ] 94 ^ 95 _ 96 ` 97 a 98 b 99 c 100 d 101 e 102 f 103 g 104 h 105 i 106 j 107 k 108 l 109 m 110 n 111 o 112 p 113 q 114 r 115 s 116 t 117 u 118 v 119 w 120 x 121 y 122 z 123 { 124 | 125 } 126 ~ 

# while loop
- If you want to appply unlimited iterations then we choose while loop

- Any loop we required 3 things

    - Intial point or start point

    - Condition to stop

    - Increment or decrement

- In for loop we are using all these 3 in a single line

- In while Loop we will use all these 3 in 3 lines



In [39]:
i=1
while i<10:
    print(i)
    i=i+1

1
2
3
4
5
6
7
8
9


In [40]:
i=1
while i!=0:
    print(i)
    i=i+1
    if i==10:
        break

1
2
3
4
5
6
7
8
9


# break
- **True**

- When we want to use infinite iterations will use True

- If we want to stop the loop, inside we will apply a condition and break

In [42]:
i=1
while True:
    print(i)
    i=i+1
    if i==10:
        break

1
2
3
4
5
6
7
8
9


In [43]:
for i in range(-10,-1):
    print(i, end=" ")

-10 -9 -8 -7 -6 -5 -4 -3 -2 

In [45]:
i=-10
while True:
    print(i, end="")
    i=i+1
    if i==0:
        break

-10-9-8-7-6-5-4-3-2-1

In [46]:
i=-10
while i<0:
    print(i, end="")
    i=i+1

-10-9-8-7-6-5-4-3-2-1

In [47]:
#wap to print -1 to -10
# using for loop and while loop
for i in range(-1,-10,-1):
    print(i,end=' ')


-1 -2 -3 -4 -5 -6 -7 -8 -9 

In [49]:
i=-1
while i>=-10:
    print(i,end="")
    i=i-1

-1-2-3-4-5-6-7-8-9-10

# string
- triple-doc stringstring
- Values can not change using Index operations called as **immutable**

- some thing can be change

- here if we want to change a prticular letter at particular index

- 'hello' :  3rd index is 'l', i want to change 'l' to 'L'

- 'heLlo'

- **by using index operation**

- **STRINGS ARE IMMUTABLE**

In [51]:
str7="""Hello how are you
       Im good"""
print(str7)

Hello how are you
       Im good


In [52]:
type("python")

str

In [53]:
print("python")

python


In [54]:
str1="python"
len(str1)

6

In [55]:
max("python")

'y'

In [56]:
min("python")

'h'

In [61]:
str1=reversed("python")
for i in str1:
    print(i,end="")

nohtyp

In [62]:
str1=sorted("python")
for i in str1:
    print(i,end="")

hnopty

In [1]:
str1="Python"
for i in str1:
    if i=="P":
        print(i)

P


In [3]:
str1=sorted("python",reverse=True)
for i in str1:
    print(i,end="")

ytponh

In [4]:
#index
str1="krishna"
print(str1[0])
print(str1[-1])

k
a


**slice**

- cut into pieces

- which means we can extract some part of string

- For that we need to undertsand the start , stop, step

- s[start:stop:step]


$pattern-1$

**str1[:stop]**

- Nothing mentioned at start place:  0

- direction nothing mentioned means postive

- step size also nothing mentioned means 1
  
- last=stop-1

In [7]:
str1="hai how are you"
# start=0
# last =(stop)-1   stop:3
#  range(0,3)  or  range(3)
# str1[start:stop]

str1[4]  # index
str1[:4] #  <start>:<stop>
# Nothing mentioned means '0'
# 0  1  2  3
# h  a  i  ''

'hai '

$pattern-2$

**str1[start:]**

In [8]:
str1="hai how are you"
str1[5:]
# start=5
# nothing mentioned means till last
# direction nothing mentioned is postive
# step  size nothing mentioned means +1

'ow are you'

In [9]:
str1="hai how are you"
str1[:]

'hai how are you'

In [10]:
str1="hai how are you"
str1[::-1]

'uoy era woh iah'

**concatenation**

In [12]:
str1='hello'
str2='python'
str1+str2
str1+' '+str2

'hello python'

In [13]:
str1='hello '
str1*3

'hello hello hello '

# string-part2
**String methods**
- you need to treat strings is a package

- so apply dir on any string we will get the methods

In [14]:
dir('')

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

$capitalize$

Return a capitalized version of the string.

More specifically, make the first character have upper case and the rest lower

In [16]:
str='python'
str.capitalize()

'Python'

In [17]:
str.upper()

'PYTHON'

In [18]:
str.lower()

'python'

$casefold$

**Return a version of the string suitable for caseless comparisons**

In [20]:
str.casefold()

'python'

$center$

In [21]:
str.center(10,'$')

'$$python$$'

In [22]:
str.center(10,'$')
# total new string width =10
# already given string len=6
# so extra 4 $ will add
# string should be in center
# 2$ are left side 2$ are right side

'$$python$$'

In [23]:
'python'.center(2,'$')

'python'

**unique vowels or items**

- always make an empty string

- iterate the letters of original string through for loop

- whenever vowel condition match keep that letter in empty string

- now onwards do the two checks

    - is that vowels

    - is that present in that new string

In [27]:
str1='aabbccddee'
str2=""
for i in str1:
    if i in "aeiou":
        if i not in str2:
            str2=str2+i
len(str2)

2

In [29]:
str1="krishna is good girl"
str1.title()

'Krishna Is Good Girl'

In [30]:
# You want count the number of 'a'  after 5th index
# You want count the number of 'a'  between 5th and 10th index
str1='ola ola ola' 
str1.count('a',7)

1

In [31]:
str1='ola ola ola' 
str1.count('a',4,21)

2

In [32]:
str1='helo'
str1.replace('l','L')

'heLo'

In [33]:
str1='hellllo'
str1.replace('l','L',2)  # count=2

'heLLllo'

In [34]:
str1='restart'
str1.replace('r','$')

'$esta$t'

$index$

- index will give the index of letter in a given string

- By default It will provide the first occurence of letter only

- If you want next occurence index of same letter you need to give start,stop

- Like we given in count method

In [36]:
str1='hai hai hai hai'
str1.find('a')
# index of 'a'
# but so many 'a' are there
# by defualt it will give first occurence index

1

**if substring not found**

- do the shift+tab understand the docstring

- if they mentioned any error or not

- for count and replace they did not mention any error

- for index error is mentioned 

- for find it is mentioned as -1

- Count  will give zero

- Replace will give same string

- index will give error

- find will give -1

In [38]:
'abc'.isalnum()

True

In [39]:
'ABC'.isupper()

True

$strip$-$lstrip$-$rstrip$

In [42]:
name1=' python '
name2='python '
name3=' python'
name1.strip(),name1.lstrip(),name1.rstrip()

('python', 'python ', ' python')

In [43]:
name2.strip(),name2.lstrip(),name2.rstrip()

('python', 'python ', 'python')

In [44]:
name3.strip(),name3.lstrip(),name3.rstrip()

('python', 'python', ' python')

In [45]:
name4=' python$$$'
name4.lstrip().rstrip('$')

'python'

In [47]:
#split()
str='hello how are you'
str.split()

['hello', 'how', 'are', 'you']

In [48]:
str1='hello how, are you'
str1.split(',')

['hello how', ' are you']

In [49]:
str2='hello how, are you'
str2.split('o')

['hell', ' h', 'w, are y', 'u']

$startswith$-$endswith$

In [53]:
str=input()
str2=str.endswith("@gmail.com")
print(str2)

 krishna@gmail.com


True


# list
- How to display

- type

- len

- max

- min

- sum

- reversed 

- sorted

- in

- not in

- index

- for loop

- mutable

- slice

- concatenation

- Methods

- List items can be heterogeous

    - Which means list items can be all data types

- List items can be duplicate

    - Which means list can have same value multiple times

- List can have another list inside: List in list

In [54]:
list1=[1,2,3,4]  # array of elements
list1

[1, 2, 3, 4]

In [55]:
type(list1)

list

In [56]:
list2=['Apple','Ball','Cat']
list2

['Apple', 'Ball', 'Cat']

In [57]:
list3=[1,2,3,'Apple','Ball','Cat']
list3

[1, 2, 3, 'Apple', 'Ball', 'Cat']

In [58]:
list6=['Apple','Ball',[1,2,3]]
list6

['Apple', 'Ball', [1, 2, 3]]

In [60]:
list1=[1,2,3,4]
list2=['Apple','Ball','Cat']
list3=[1,2,3,'Apple','Ball','Cat']
list4=[1,'Apple',10.5,True]
list5=[100,100,100]
list6=['Apple','Ball',[1,2,3]]

In [61]:
len(list1)

4

In [62]:
max(list1)

4

In [63]:
min(list1)

1

In [65]:
val=reversed(list1)
for i in val:
    print(i)

4
3
2
1


In [66]:
sorted(list1,reverse=True)

[4, 3, 2, 1]

In [67]:
33 not in list1

True

In [68]:
3 in list1

True

In [69]:
list1=[10,20,30,40,50,60]
list1[0],list1[1],list1[2]

(10, 20, 30)

In [70]:
list1=[1,2,3,4]
for i in range(len(list1)):
    print(list1[i])

1
2
3
4


In [71]:
list1=[1,2,3,[10]]
list1[3][0]

10

In [72]:
list1=[1,2,3,['A',['B']]]
len(list1)

4

In [73]:
list1=[1,2,3,['Apple',['Fruites',['Cherry']]]]
len(list1[3])
list1[3][1][1][0]

'Cherry'

In [75]:
list1=[1,2,3,['Apple',['Fruites',['Cherry']]]]
len(list1)

4

In [76]:
list1=[[[1,2,3,['Apple',['Fruites',['Mango',['Cherry']]]]]]]
len(list1)

1

In [77]:
for i in '5x*x+6x+c':
    if i.isdigit():
        print(i)

5
6


**Mutable - Immutable**
- Python mutable concept always based on index

- list are mutable means we can change the values using index

In [78]:
l=[1,2,3,4]
l[0]=100
l

[100, 2, 3, 4]

In [79]:
l1=[1,2,3,4]
l2=[10,20,30,40]
l1+l2

[1, 2, 3, 4, 10, 20, 30, 40]

In [80]:
l1*2

[1, 2, 3, 4, 1, 2, 3, 4]

In [81]:
dir([])

['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

**append**

- append means adding a number at the end of the list

- append is a very basic method we will use many times in our code

- append will use mainly to store the result


In [83]:
number=['one','two','three','four','five']
number['four']

TypeError: list indices must be integers or slices, not str

In [84]:
l=[1,2,3,4]
l.append(100)
l

[1, 2, 3, 4, 100]

In [85]:
l.append(100)
l

[1, 2, 3, 4, 100, 100]

In [86]:
l=[]
l.append(10)
l.append('apple')
l

[10, 'apple']

In [87]:
l=[]
for i in range(1,6):
    l.append(i*i)

In [88]:
l

[1, 4, 9, 16, 25]

In [90]:
# Q2) l=['Hyd','blr','Chennai','pune']
# out=['Hyd','Chennai']

# is title
# Step-1: make one empty list
# Step-2: Iterate each word from list
# Step-3: apply is title method if condition
# Step-3: if that condition is satisfy
# Step-5: then append the result in empty list
l=['Hyd','blr','Chennai','pune']
out=[]
for i in l:
    if i.istitle():
        out.append(i)
out

['Hyd', 'Chennai']

In [92]:
#Q3) l=['HYD','blr','Chennai','PUNE']
# out=['HYD','PUNE']
l=['HYD','blr','Chennai','PUNE']
out=[]
for i in l:
    if i.isupper():
        out.append(i)

out

['HYD', 'PUNE']

In [102]:
s='can canner can not a can but canner can you make a can'
s1=s.split()
for i in s1:
    count=0
    for j in s1:
        if i==j:
            count=count+1
        print(i,count)
    

can 1
can 1
can 2
can 2
can 2
can 3
can 3
can 3
can 4
can 4
can 4
can 4
can 5
canner 0
canner 1
canner 1
canner 1
canner 1
canner 1
canner 1
canner 2
canner 2
canner 2
canner 2
canner 2
canner 2
can 1
can 1
can 2
can 2
can 2
can 3
can 3
can 3
can 4
can 4
can 4
can 4
can 5
not 0
not 0
not 0
not 1
not 1
not 1
not 1
not 1
not 1
not 1
not 1
not 1
not 1
a 0
a 0
a 0
a 0
a 1
a 1
a 1
a 1
a 1
a 1
a 1
a 2
a 2
can 1
can 1
can 2
can 2
can 2
can 3
can 3
can 3
can 4
can 4
can 4
can 4
can 5
but 0
but 0
but 0
but 0
but 0
but 0
but 1
but 1
but 1
but 1
but 1
but 1
but 1
canner 0
canner 1
canner 1
canner 1
canner 1
canner 1
canner 1
canner 2
canner 2
canner 2
canner 2
canner 2
canner 2
can 1
can 1
can 2
can 2
can 2
can 3
can 3
can 3
can 4
can 4
can 4
can 4
can 5
you 0
you 0
you 0
you 0
you 0
you 0
you 0
you 0
you 0
you 1
you 1
you 1
you 1
make 0
make 0
make 0
make 0
make 0
make 0
make 0
make 0
make 0
make 0
make 1
make 1
make 1
a 0
a 0
a 0
a 0
a 1
a 1
a 1
a 1
a 1
a 1
a 1
a 2
a 2
can 1
can 1
can 2
can 2
c

In [103]:
l=[1,2,3]
for i in l:
    for j in l:
        print(i,j)

1 1
1 2
1 3
2 1
2 2
2 3
3 1
3 2
3 3


In [104]:
# WAP ask the user say 5 times hello
# save hello in a list 
# use list comprehension
['hello' for i in range(5)]

['hello', 'hello', 'hello', 'hello', 'hello']

In [112]:
l1=[1,2,3,4]
l2=[10,20,30,40]
out=[]
for i in range(len(l1)):
      out.append(l1[i]+l2[i])
out

[11, 22, 33, 44]

**List comprehensions**
- list comprehension will not use append

- inside empty list only we are writing the co
- list comprehension will not use append

- inside empty list only we are writing the codede

In [114]:
# q1) wap ask the user print 1,5 numbers square
#    save the result in a list
l=[]
for i in range(1,6):
    l.append(i*i)
l

[1, 4, 9, 16, 25]

In [115]:
#pattern-1
# WAP ask the user say 5 times hello
# save hello in a list 
# use list comprehension
['hello' for i in range(5)]

['hello', 'hello', 'hello', 'hello', 'hello']

In [117]:
l=['Hyd','blr','Chennai','pune']
out=[i for i in l if i.istitle()]
out

['Hyd', 'Chennai']

In [119]:
#pattern-2
# Q3) l=['HYD','blr','Chennai','PUNE']
# out=['HYD','PUNE']
l=['HYD','blr','Chennai','PUNE']
out=[i for i in l if i.isupper()]
out

['HYD', 'PUNE']

$pattern-3$

**for-if-else**

In [120]:
# **for-if-else**
# [<output> <for loop> ]
# [<output> <for loop>  <if condition>]
# [<if_output> <if condition> <else> <else_output> <forloop>]

# out=[]
# for i in range(1,11):
#     if i%2==0:
#         out.append(f"even {i}")
#     else:
#         out.append(f"odd {i}")
# out

[f"even {i}" if i%2==0 else f"odd {i}" for i in range(1,11)]

['odd 1',
 'even 2',
 'odd 3',
 'even 4',
 'odd 5',
 'even 6',
 'odd 7',
 'even 8',
 'odd 9',
 'even 10']

In [121]:
l=[10,10,10,100,200,300]
l.count(10)

3

**unique vowels wrapper**

In [122]:
s='can canner can not a can but canner can you make a can'
l=s.split() # ['can','canner']
l1=[]
for i in l:
    if i not in l1:
        print(i,l.count(i))
        l1.append(i)

can 5
canner 2
not 1
a 2
but 1
you 1
make 1


**Extend**
**difference between extend and concatenation**

- extend will overwrite the output

- concatenation will  same as it is

In [124]:
l1=[1,2,3]
l2=['a','b','c']
l1.extend(l2)
l1

[1, 2, 3, 'a', 'b', 'c']

In [125]:
l1=[1,2,3]
l2=['a','b','c']
l1.extend(l2) # [1,2,3,a,b,c]
l1.append(l2) # [1,2,3,[a,b,c]]
l1+l2         # [1,2,3,a,b,c]

[1, 2, 3, 'a', 'b', 'c', ['a', 'b', 'c'], 'a', 'b', 'c']

In [126]:
l=[1,2,3,4]
l.insert(3,100)
l

# at second index i want insert 100

[1, 2, 3, 100, 4]

In [128]:
#concatenation
# extend
# append
# insert

In [129]:
l=[i for i in range(1,20,3)]
l.insert(4,'apple')

**pop-remove**
- pop will remove the last value

- inside pop we have index argument

- by default index=-1  , -1 means last value

- pop will remove the value , the output will overwrite

- pop also return the item value, which it is removing
- remove will take the direct item name

- pop will take only the index

- both raise the error

In [131]:
l=[1,2,30,4]
id=l.index(30)
print(id)
l.pop(2)
l

2


[1, 2, 4]

In [132]:
l=['apple','ball','cat',10,20,'Dog','apple']
l.remove('apple')
l

['ball', 'cat', 10, 20, 'Dog', 'apple']

- append

- extend

- insert

- clear /copy

- pop/remove

- sort/reverse

- index

- count

In [134]:
sorted([10,2,3,4])

[2, 3, 4, 10]

In [135]:
[10,2,3,4].sort()

In [136]:
l=[1,2,3,4]
del(l[0])

**zip**

In [138]:
l1=[1,2,3]
l2=[10,12,13]
# apply the zip
# zip is a iterator
# get the for loop
# two values we zipped
# we need two varaibles
# for i,j in zip(l1,l2):
#     print(i+j)

out=[i+j for i,j in zip(l1,l2)]
out

[11, 14, 16]

# Dictionary

- Dictionary is represents with curly braces

- strings using quotes

- list using square brackets

- type using parnathses

- Dictionary is alway a pair key and value
- **keys are important**

- Keys have only latest value

- No duplicate keys

- Keys should be only one 

- Values can be duplicate 

- which means different keys can have same value

- List can not be a key

- tupl

- min

- max

- len

- sorted

- reveresd
- **Keys are important**

- Dictionary is a **key-value** pair

- There is no concept of index as numbers

- If you want to retrive any value , we need to provide key as index

- we can access values using only corresponding **keys**e can be a key

In [152]:
name=['Manish','Shubam','Govindh']
age=[25,20,15]
for i,j in zip(name,age):
    print(f"the {i} age is {j}")

the Manish age is 25
the Shubam age is 20
the Govindh age is 15


In [141]:
dict1={'Manish':25,'Shubam':20,'Govindh':15}
dict1

{'Manish': 25, 'Shubam': 20, 'Govindh': 15}

In [142]:
type(dict1)

dict

In [143]:
d1={'Manish':25,'Shubam':20,'Govindh':15}
d2={'Manish':'25','Shubam':'20','Govindh':'15'} 
d3={25:'Manish',20:'Shubam',15:'Govindh'} 
d4={25:50} 
d5={'Fruites':['Apple','Cherry']} 
d6={['Apple','Cherry']:'Fruites'} 
d7={'Ajay':25,'Ajay':30}
d8={25:'Ajay',30:'Ajay'} 

TypeError: unhashable type: 'list'

In [144]:
d1={'Manish':25,'Shubam':20,'Govindh':15}
d1

{'Manish': 25, 'Shubam': 20, 'Govindh': 15}

In [145]:
d2={'Manish':'25','Shubam':'20','Govindh':'15'}
d2

{'Manish': '25', 'Shubam': '20', 'Govindh': '15'}

In [146]:
d3={25:'Manish',20:'Shubam',15:'Govindh'}
d3

{25: 'Manish', 20: 'Shubam', 15: 'Govindh'}

In [147]:
d4={25:50} 
d4

{25: 50}

In [148]:
d5={'Fruites':['Apple','Cherry']} 
d5

{'Fruites': ['Apple', 'Cherry']}

In [149]:
d6={['Apple','Cherry']:'Fruites'}
d6

TypeError: unhashable type: 'list'

In [150]:
d7={('Apple','Cherry'):'Fruites'}
d7

{('Apple', 'Cherry'): 'Fruites'}

In [151]:
d9={25:'Ajay',30:
    {'Fruites':['Apple']}
   } 
d9

{25: 'Ajay', 30: {'Fruites': ['Apple']}}

In [154]:
d1={'Manish':25,'Shubam':20,'Govindh':15}
d1['Manish']

25

In [155]:
d1={'Manish':25,'Shubam':20,'Govindh':15}
print(d1['Manish'])
print(d1['Shubam'])
print(d1['Govindh'])

print(d1[i])

25
20
15
15


In [156]:
for i in d1:
    print(i,d1[i])

Manish 25
Shubam 20
Govindh 15


In [157]:
d1={'Fruites':['Apple']}
d1['Fruites'][0]

'Apple'

In [158]:
len(d1)

1

In [159]:
d1={'Fruites':{"Apple":['Kashmir']}}
d1

{'Fruites': {'Apple': ['Kashmir']}}

In [160]:
d1['Fruites']['Apple'][0]

'Kashmir'

In [162]:
#empty dict
d={}
d['Ramesh']=50
d['Suresh']=100
d['Sathish']=150
d['Manish']=200
d

{'Ramesh': 50, 'Suresh': 100, 'Sathish': 150, 'Manish': 200}

**How to create empty dictionary**

- it is very important to know how to make empty list

- how to make empty string

- how to make empty dictionary

In [163]:
str1='python'
str2=''
for i in str1:
    str2=str2+i      # str2+=i
str2

'python'

In [164]:
l1=[1,2,3,4]
l2=[]
for i in l1:
    l2.append(i)

l2

[1, 2, 3, 4]

In [165]:
d={}
d['Ramesh']=50
d['Suresh']=100
d['Sathish']=150
d['Manish']=200
d

{'Ramesh': 50, 'Suresh': 100, 'Sathish': 150, 'Manish': 200}

**Convert string to list**

In [167]:
# split: convert string to list
str1='hai how are you'
l=str1.split()
l

['hai', 'how', 'are', 'you']

**join: convert list to string**

In [168]:
l=['hai', 'how', 'are', 'you']
new_str=' '.join(l)
new_str

'hai how are you'

In [169]:
name=['Ramesh','Suresh','Manish']
age=[25,30,35]
d={}
for i,j in zip(name,age):
    d[i]=j

d

{'Ramesh': 25, 'Suresh': 30, 'Manish': 35}

**items-keys-values**
- items means both keys and values come together

- keys means only dictionary keys

- values means only dictionary values

In [171]:
d={'Ramesh': 50, 'Suresh': 100, 'Sathish': 150, 'Manish': 200}
it=d.items()
it


dict_items([('Ramesh', 50), ('Suresh', 100), ('Sathish', 150), ('Manish', 200)])

In [172]:
d={'Ramesh': 50, 'Suresh': 100, 'Sathish': 150, 'Manish': 200}
it=d.items() # items
items_list=list(it) # list type
name=[]
age=[]
for i in items_list:
    name.append(i[0])
    age.append(i[1])
name

['Ramesh', 'Suresh', 'Sathish', 'Manish']

# Lambda function
- Lambda function also a function 

- It is a kind of a single line function

- we already seen list comprehension which means single line

- lambda functions give a flexibility to stop writing many lines

- on iterative execution it will save the timing

In [174]:
def summ(num):
    return(num+10)

summ(20)

30

$pattern-1$
- function name

- variable name

- return output


In [175]:
#<function_name>= lambda <variable_name>: <return_output>

In [176]:
summ=lambda num : num+10
summ(30)

40

In [177]:
def cube(x):
    return(x*x*x)
cube(10)

1000

In [178]:
cube=lambda x:x*x*x
cube(20)

8000

$pattern-2$

**Two arguments**

In [179]:
def add(n1,n2):
    return(n1+n2)

add(10,20)

30

In [180]:
#<function name>=lambda <arg1>,<arg2>: <output>

In [181]:
add=lambda n1,n2:n1+n2
add(10,20)

30

$pattern-3$

**Default arguments**

In [182]:
avg=lambda n1,n2,n3=100 : round((n1+n2+n3)/3,2)
avg(10,20)

43.33

$Patterns-4$

**if**

In [183]:
def greet(n1):
    if n1>10:
        return("hello")
greet(20)

'hello'

In [184]:
greet= lambda n1 : 'hello' if n1>10 else 'bye'
greet(10)

'bye'

In [185]:
# wap lambda function to identidy the greatet number between two numbers

maxx=lambda n1,n2: f"{n2} is greater" if n2>n1 else f"{n1} is greater"
maxx(10,20)

'20 is greater'

$Pattern-5$

**using list**

In [186]:
list1=['hyd','blr','chennai']
# op=list2=['Hyd','Blr','Chennai']
list2=[]
for i in list1:
    list2.append(i.capitalize())

list2

['Hyd', 'Blr', 'Chennai']

In [187]:
# lambda <variable name> : <output>,<iterable>
lambda i : i.capitalize(),list1

(<function __main__.<lambda>(i)>, ['hyd', 'blr', 'chennai'])

**map**

In [189]:
list(map(lambda i : i.capitalize(),list1))

['Hyd', 'Blr', 'Chennai']

In [190]:
list1 = ['hyd','blr','chennai']
print(list(map(lambda i: i.capitalize(), list1)))
print(list(map(lambda i: i.upper(), list1)))
print(list(map(lambda i: i.lower(), list1)))
print(list(map(lambda i: i.center(10,"*"), list1)))

['Hyd', 'Blr', 'Chennai']
['HYD', 'BLR', 'CHENNAI']
['hyd', 'blr', 'chennai']
['***hyd****', '***blr****', '*chennai**']


In [191]:
# step-1: whenever you are applying a condition on list
#         write only condition with out using if
lambda i: '#' in i,list1

list(map(lambda i: '#' in i,list1))

[False, False, False]

**filter method**

In [192]:
list(filter(lambda i: '#' in i,list1))

[]

In [193]:
list1 = ['hyd','blr','chennai','mumbai']
#list2=['hyd','blr'] # len(i)<=3

list(filter(lambda i: len(i)<=3,list1))

['hyd', 'blr']

# Text book

**Def. of Module:**
- A module is a collection Functions, Variables and Class names 
Examples: math, cmath, calendar, rand..e
- that Garbage Collector 
Collects Un-used Memory space and Improves peformanace of python Real Time Applications
- static Typed Programming Languages:declare variables mandatory:java,c#,c
- Dynamically Typ:ed Programming Languages:not necessary to use Variable declration:python
- In Python Programming, All Values are stored in the form of OBJECTS and Behind of object 
there exist CLASS
- The process of converting Technical Error Messages into user-Friendly Error message is called Exception Handling
Literal-value passing as input to the program : Int,float,string,boolean,date
- all variables are called objects,literal stored in distinct names :variables
- All class Name can be used as Variable Name bcoz class Names are not keywortc

**Datatypes**
- 1.Fundamental Category Data types:store single value
- int
  - Decimal:0,1,2,3,4,5,6,7,8,9
      - Base:10
  - Binary:0,1
      - Base:2
      - 0b or 0B
  - Octal:0,1,2,3,4,5,6,7
      - Base:8
      - 0o or 0O
  - Hexa Decimal:0,1,2,3,4,5,6,7,8,9,A,B,C,D,E,F
      - Base:8
      - 0x or 0X
- float:only allow decimal,not allow binary,octal,hexa
  - example:a=2e3
     - 2000.0
  -a=10e-2
     - 0.1
- boolean:True or False
- complex:a+bj:real+imginary
  - the real and imginary parts are by default belogs to float
- 2.Sequence Category Data types
- str:pre-defined class:quotes""
  - multi_linedata:"""    """
  - operations
      - 1.Indexing
      - 2.Slicing     
- bytes
- bytearray
- range